.NET SDK managed workload installer

sfoslund opened the pull request:
Not included in this PR:
- ~~Garbage collection~~
- Manifest updates

Testing notes:
- Any workloads with all packs on `https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json` will successfully install (ex macOS, iOS workloads), others won't until we expand the nuget package installer API
- Example test command: `dotnet workload install microsoft-ios-sdk-full --skip-manifest-update`, make sure to include skip manifest update flag

sfoslund committed:
Add workload install orchestration logic

sfoslund committed:
Adding .NET SDK managed workload resolver

sfoslund committed:
Merge remote-tracking branch 'upstream/main' into WorkloadInstall

sfoslund committed:
Cleaning up merge

joeloff commented on src/Resolvers/Microsoft.DotNet.NativeWrapper/EnvironmentProvider.cs:
```diff
@@ -89,5 +89,51 @@ public static string GetDotnetExeDirectory(Func<string, string> getEnvironmentVa
+            var tmp = _getEnvironmentVariable(Interop.RunningOnWindows ? "TEMP" : "TMPDIR");
+            if (string.IsNullOrEmpty(tmp))
+            {
+                throw new Exception("Unable to resolve user temp");
```
Should these messages be localized or do users never see them?

sfoslund replied:
I think we generally don't localize error messages since theoretically the user won't ever see them (if everything goes right :) )

joeloff replied:
Hmm, not buying that completely. If users don't see errors, then there won't be a point in reporting a message :) It can impact accessibility, e.g. screen readers that assist non-English speaking users.

sfoslund committed:
Moving workload installation records from sdk-manifest folder to metadata folder

joeloff commented on src/Resolvers/Microsoft.DotNet.NativeWrapper/EnvironmentProvider.cs:
```diff
@@ -89,5 +89,51 @@ public static string GetDotnetExeDirectory(Func<string, string> getEnvironmentVa
+
+        public string GetTempDirectory()
+        {
+            var tmp = _getEnvironmentVariable(Interop.RunningOnWindows ? "TEMP" : "TMPDIR");
```
Why are we using environment variables? Path.GetTempPath works on both Windows and Linux, and it implements a set of fallback rules. For example, on Linux it falls back to /tmp if TMPDIR doesn't exist, but here we would throw an exception.

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadUnitInstaller.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal abstract class IWorkloadUnitInstaller : IWorkloadInstaller
```
Should this be named WorkloadUnitInstallerBase since it's abstract and not an interface?

sfoslund committed:
Adding garbage collection

wli3 commented:
Sorry please give me sometime to review this change carefully. including run the tests and see the test coverage

wli3 commented on src/Cli/dotnet/NugetPackageInstaller/NuGetPackagePathResolver.cs:
```diff
@@ -0,0 +1,25 @@
+            return string.Empty; // Extract package directly to target directory
+        }
+
+        public override string GetPackageFileName(PackageIdentity packageIdentity)
```
Why this need an override. The base class is https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.Packaging/PackagePathResolver.cs#L57
```
public virtual string GetPackageFileName(PackageIdentity packageIdentity)
        {
            var fileNameBase = GetPathBase(packageIdentity);

            fileNameBase.Append(PackagingCoreConstants.NupkgExtension);

            return fileNameBase.ToString();
        }
```

They do not do the same?

sfoslund replied:
The base class adds some extra directories to the target path. I overrode it so we could specify the directory where the nupkg would go directly.

sfoslund replied:
Yes, exactly, I'll add a comment.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadInstaller.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal interface IWorkloadInstaller
```
@dsplaisted please review this interface. "WorkloadInstaller" is not a concept in the spec. And the name is too vague. We should come up with a better naming and coin that in the spec

sfoslund replied:
This was meant to act as the IAL but I figured that wasn't a very descriptive name in this context. Definitely open to other suggestions, though.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadInstaller.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        InstallationUnit GetInstallationUnit();
+
+        void InstallWorkloadManifest(string manifestId, string manifestVersion, string sdkFeatureBand);
```
http://wiki.c2.com/?PrimitiveObsession
now C# have record. it is easier now. We should warp all these strings to a type, even if there is only one thing. If we worry about the perf, we can make them structs like packageid.

manifestId, manifestVersion, sdkFeatureBand, workloadId should all be separate types. Some of them may make sense to combine later when the need of abstraction is clear. Also, there should be only one way to present these IDs in our system. For example, manifestId can only be all lowercased. sdk feature band can only be a number or 5xx. We need to do validation, normalization when they come in to the system from domain boundary (parser etc...). I did the similar thing to packageid.

wli3 reviewed (commented):
not done yet

joeloff commented on src/Cli/dotnet/NugetPackageInstaller/NuGetPackageInstaller.cs:
```diff
@@ -15,23 +15,24 @@
             var cancellationToken = CancellationToken.None;
             var cache = new SourceCacheContext() { DirectDownload = true, NoCache = true };
-            var source = Repository.Factory.GetCoreV3(sourceUrl);
+            var source = Repository.Factory.GetCoreV3(_sourceUrl);
```
Are we just deferring all the URL checks to NuGet, e.g. if it's actually a V3 source or even a valid URL

sfoslund replied:
For now we're only using the xamarin source, but Will is going to expand this later to be more adaptable.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal class WorkloadInstallManager
```
https://blog.codinghorror.com/i-shall-call-it-somethingmanager/

we should give it a better name and put it in spec. We have workload install, workload install manager(aka the caller of "workload install"), it is a reflection of we don't have a clear cut on the abstraction.

dsplaisted replied:
Could this logic just be moved into the `WorkloadInstallCommand` class?  Then we wouldn't need a separate "manager".

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+    internal class WorkloadInstallManager
+    {
+        private readonly IReporter _reporter;
+        private IWorkloadInstaller _workloadInstaller;
```
nit: mark them readonly when possible

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+            _reporter.WriteLine();
+        }
+
+        private void InstallWorkloadComponents(IEnumerable<string> workloadIds, string featureBand)
```
we don't have the concept of "component" in the spec(if we have a spec, code should align wtih the spec). Just call it "install packs and add record". The abstraction of "component" for packs and record is not too useful.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallerFactory.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    internal class WorkloadInstallerFactory
+    {
+        public static IWorkloadInstaller GetWorkloadInstaller(IReporter reporter)
```
same here, if constructor works, use constructor

sfoslund replied:
I added this factory for when we have other types of workload installers. For now, since this the managed installer is the only one implemented, I wasn't sure if we wanted to add logic to error here if the SDK managed installer isn't appropriate (for example, with global installs that need MSIs)

joeloff replied:
We would eventually need something to decide which installer to instantiate. Wouldn't hurt to have something here that we can iterate over as new installers become available.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+        {
+            if (useOfflineCache)
+            {
+                throw new NotImplementedException();
```
make sure todo and NotImplementedException is refected in the github issues. So we don't miss them later

sfoslund replied:
Sure, I'll keep https://github.com/dotnet/sdk/issues/16483 up to date with these items

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+                        }
+                        else
+                        {
+                            _reporter.WriteLine(string.Format(LocalizableStrings.WorkloadPackAlreadyInstalledMessage, packInfo.Id, packInfo.Version));
```
test for failure case

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+            return workloadResolver.TryGetPackInfo(packId);
+        }
+
+        protected virtual WorkloadResolver GetWorkloadResolver(string featureBand)
```
constructor injection is not enough?

usually we do this for legacy code when the separation of functionality is not intentional

sfoslund replied:
Yeah, that's something I wanted to check, probably with @dsplaisted. The workload resolvers are by feature band (in that they load and work off of only the manifests in the feature band we specify), right? If so, then we'll have to create one for each feature band when we garbage collect here: https://github.com/dotnet/sdk/pull/16530/files#diff-c01acc54f7e26514dc3843cef3eb6b1418943a3e139ea89abbaa27af964dc39bR128

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+                            if (packInfo.Kind.Equals(WorkloadPackKind.Library) || packInfo.Kind.Equals(WorkloadPackKind.Template))
+                            {
+                                Directory.CreateDirectory(Path.GetDirectoryName(packInfo.Path));
+                                File.Copy(packPath, packInfo.Path);
```
As we discussed in the meeting, a more robust way is to put everything in a temp directory first and then run a FileAccessRetrier.RetryOnMoveAccessFailure(() => Directory.Move

If the operation fails in the halfway. The folder will be less of a headache (be there as a whole or not at all)

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+            {
+                if (packInfo.Kind.Equals(WorkloadPackKind.Library) || packInfo.Kind.Equals(WorkloadPackKind.Template))
+                {
+                    File.Delete(packInfo.Path);
```
check for existence?

wli3 replied:
It will never happen or there is a missing test?

sfoslund replied:
The existence check is done in `PackIsInstalled`, so at this point we know this file exists

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,283 @@
+                }
+                else
+                {
+                    Directory.Delete(packInfo.Path, true);
```
Same here

wli3 reviewed (commented):
Finally finished a round of check.

Awesome work as always!

sfoslund commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:
```diff
@@ -80,7 +80,10 @@ public IEnumerable<string> GetManifestDirectories()
                     foreach (var workloadManifestDirectory in Directory.EnumerateDirectories(_manifestDirectories[0]))
                     {
-                        yield return workloadManifestDirectory;
+                        if (!Path.GetFileName(workloadManifestDirectory).StartsWith("."))
```
Oops, that's leftover from a earlier iteration, thanks for pointing it out!

sfoslund committed:
Converting primitive types to workload types

sfoslund committed:
PR feedback

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/install/IWorkloadInstaller.cs:
```diff
@@ -9,15 +9,15 @@ internal interface IWorkloadInstaller
         InstallationUnit GetInstallationUnit();
 
-        void InstallWorkloadManifest(string manifestId, string manifestVersion, string sdkFeatureBand);
+        void InstallWorkloadManifest(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand sdkFeatureBand);
```
@wli3 Does this mean the manifest version is now a proper version? Last time I checked it was defined as a long.

sfoslund committed:
Fixing garbage collection bug

sfoslund commented:
@dsplaisted @wli3 @joeloff I think I'm caught up on most of the feedback, is there anything else I should look at?

I'm also consistently getting a test failure that I can't repro anywhere that's caused by a `System.UnauthorizedAccessException` error when looking at a directory that I lay down in the test. Have any of you seen this kind of things before? https://dev.azure.com/dnceng/public/_build/results?buildId=1074516&view=ms.vss-test-web.build-test-results-tab&runId=32985386&resultId=100738&paneView=debug

wli3 commented:
Let me check again

sfoslund committed:
Fixing workload managed install garbage collection bug

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+            }
+
+            _reporter.WriteLine();
+            _reporter.WriteLine(string.Format(LocalizableStrings.InstallationSucceeded, string.Join(", ", workloadIds)));
```
I'm not sure if joining items with commas is localization-neutral.  I'm not sure if there's a better pattern.

wli3 replied:
It is good enough. Maybe join by whitespace. But so far that's how existing code handle it.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+                var installer = _workloadInstaller as PackWorkloadInstallerBase;
+
+                var workloadPacksMap = workloadIds
+                    .Select(workloadId => (workloadId, _workloadResolver.GetPacksInWorkload(workloadId.ToString()).Select(packId => _workloadResolver.TryGetPackInfo(packId))));
```
Different workloads may use the same packs.  I would create a de-duplicated list of workloads packs to install here.  They don't need to be linked to which workload(s) they were installed for.  Then just iterate separately over the list of packs to install and workloads to write installation records for.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallManager.cs:
```diff
@@ -0,0 +1,100 @@
+                        {
+                            foreach (var packId in packsToInstall)
+                            {
+                                installer.RollBackWorkloadPackInstall(packId, sdkFeatureBand);
```
I think there's an issue with this rollback logic.  A workload pack might already have been installed for another workload that was already installed, but it looks like if we were to roll back the install of a workload that also depended on the workload pack, we'd end up deleting the workload pack and breaking the previously installed workload.

sfoslund replied:
I noticed this bug in too and changed it in my latest commit, now I'm checking to make sure that there aren't any pack install records left before uninstalling the pack.

dsplaisted reviewed (approved):
Great job!  Feel free to go ahead and merge and address the feedback separately if that works better for you.

I didn't review the tests yet, I will try to still do that.

sfoslund commented on src/Cli/dotnet/commands/dotnet-workload/install/ManifestVersion.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal struct ManifestVersion : IEquatable<ManifestVersion>, IComparable<ManifestVersion>
```
@wli3 suggested it in previous feedback, to avoid using primitives for all these concepts

sfoslund commented on src/Cli/dotnet/commands/dotnet-workload/install/ManifestId.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal struct ManifestId : IEquatable<ManifestId>, IComparable<ManifestId>
```
We convert ids to lower case in the constructor so I don't think it should make a difference.

sfoslund commented on src/Cli/dotnet/commands/dotnet-workload/install/PackWorkloadInstallerBase.cs:
```diff
@@ -0,0 +1,36 @@
+        // Workload pack installer unique methods:
+        public abstract void InstallWorkloadPack(PackInfo packInfo, SdkFeatureBand sdkFeatureBand, bool useOfflineCache = false);
+
+        public abstract void RollBackWorkloadPackInstall(PackInfo packInfo, SdkFeatureBand sdkFeatureBand);
```
I patterned this off of the tools rollback logic (which I assume @wli3 implemented). Are there any examples of the rollback pattern you suggest in our code base that I can take a look at?

wli3 replied:
@dsplaisted it is no that smart. You need to implement yourself. Transaction scope will call roll back at the right time.

wli3 commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs:
```diff
@@ -0,0 +1,307 @@
+                    {
+                        File.Delete(file);
+                        var packagePath = Path.GetDirectoryName(file);
+                        while (!Directory.EnumerateFileSystemEntries(packagePath).Any())
```
just note deleting empty directories will cause concurrency issue, if we care about concurrency.

sfoslund committed:
PR feedback

sfoslund commented:
@dsplaisted thanks for the feedback! I think I addressed all of it and there were some significant changes (especially in garbage collection) in the last commit, do you want to take a look before I merge?

The only piece I've left out of this PR is refactoring this into a separate project per [this comment](https://github.com/dotnet/sdk/pull/16530#discussion_r608237834). I've added it to https://github.com/dotnet/sdk/issues/16483 as a follow up task.

sfoslund committed:
Add workaround for testing workload commands without env var enabled

sfoslund commented:
@wli3 I had to add a workaround for testing workload commands when the `DEVENABLEWORKLOADCOMMAND` environment variable isn't set, is there somewhere you're tracking removing that env var so I can add a note to remove 35900a5 as well?

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        [Fact]
+        public void GivenManagedInstallItCanGetFeatureBands()
```
I think `GetFeatureBandsWithInstallationRecords` should be changed so it only returns bands where there are actually workload installation records.  So this test should be updated to account for that, and to test both empty and non-empty `InstalledWorkloads` directories.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        [Fact]
+        public void GivenManagedInstallItCanGetInstalledWorkloads()
```
For this test I'd suggest writing some different workloads for a different feature band to verify it doesn't pick those up.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        [Fact]
+        public void GivenManagedInstallItDetectInstalledPacks()
```
Grammar nit

```suggestion
        public void GivenManagedInstallItDetectsInstalledPacks()
```

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+
+            installer.InstallWorkloadPack(packInfo, new SdkFeatureBand(version));
+
+            _reporter.Lines.Should().Contain(string.Format(LocalizableStrings.WorkloadPackAlreadyInstalledMessage, packInfo.Id, packInfo.Version));
```
Instead of (or in addition to) verifying the message, could this test check that the NuGet package installer was never called to download the package?

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+                installer.InstallWorkloadPack(packInfo, new SdkFeatureBand(version));
+
+                // Install should have failed
+                true.Should().BeFalse();
```
Switch try/catch and `true.Should().BeFalse()` for `Assert.Throws`.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        [Fact]
+        public void GivenManagedInstallItCanRollBackInstallFailures()
```
Could we have coverage of what happens when multiple packs are supposed to be installed, and some install successfully but then one fails and they all need to be rolled back?

sfoslund replied:
That's what this test covers: https://github.com/dotnet/sdk/pull/16530/files#diff-c8bc89094b206b506f35f0cb9ac6e55591ae2468130efe0a19dab1810c0f0c1cR52, the first pack installs successfully, the second fails, and they are both rolled back.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenNetSdkManagedWorkloadInstall.cs:
```diff
@@ -0,0 +1,255 @@
+        }
+
+        [Fact]
+        public void GivenManagedInstallItDoesNotRemovePacksWithInstallRecords()
```
Can we add some packs to this test that aren't part of the `xamarin-android-build` workload and verify that they get garbage collected?

sfoslund commented:
Thanks for the feedback, I'll address it in a follow up PR
